File: lib/debruijn/word.rb
class Word
  attr_reader :letters
  def initialize(*letters)
    @letters = letters.flatten
  end

  def append(letter)
    return Word.new(Array.new(@letters) << letter)
  end

  def pipe(letter)
    result = Array.new(@letters).push(letter)
    result.shift
    Word.new(result)
  end

  def subwords(k)
    copy = [].concat(@letters).concat(@letters)
    (0..@letters.length).each do |index|
      yield Word.new(copy[index..(index + k - 1)])
    end
  end

  def ==(word) 
    self.eql?(word)
  end

  def eql?(word)
    self.class.eql?(word.class) && self.letters.eql?(word.letters)    
  end

  def hash
    self.letters.hash
  end

  def to_s
    "(#{@letters.join(', ')})"
  end
end

File: debruijn/combinatorics/word.py
class Word:
    def __init__(self, *letters):
        self.letters = list(letters)
        
    def append(self, letter):
        letters = self.letters[:]
        letters.append(letter)
        return Word(*letters)

    def pipe(self, letter):
        letters = self.letters[1:]
        letters.append(letter)
        return Word(*letters)
    
    def __eq__(self, other):
        if (isinstance(other, Word)):
            return self.letters == other.letters
        return False

    def __hash__(self):
        return hash(self.letters)
    
    def __repr__(self):
        return "Word({0})".format(",".join(self.letters))
    
    def __str__(self):
        return "[ {0} ]".format(", ".join(self.letters))
